Fixed #29232 -- Added title option for admin filter when providing field name

dsanders11 opened the pull request:
https://code.djangoproject.com/ticket/29232

dsanders11 commented:
This PR also includes a change to the admin `checks.py` file to fix cases where `issubclass` would raise a `TypeError` during the checks which would be visible to the user instead of the intended check error message. This can be pulled out into a separate PR if desired.

carltongibson commented:
> This can be pulled out into a separate PR if desired.

Yes please.

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -114,9 +114,17 @@ def get_filters(self, request):
+                        title = second_param
+                    else:
+                        # This is a custom FieldListFilter class for a given field.
+                        field_list_filter_class = second_param
```
I'd write this block a little differently: 

```python
if isinstance(list_filter, (tuple, list)):
    # This is a custom FieldListFilter class for a given field.
    field, field_list_filter_class = list_filter
    if isinstance(field_list_filter_class, str):
        title = field_list_filter_class
        field_list_filter_class = FieldListFilter.create
else:
    ...
```

dsanders11 replied:
I'm going to respectfully disagree here, but if you insist I'll rewrite it in this style.

IMO in your version the comment is incorrect (it's a custom `FieldListFilter` class **or** a `str`), and I think it's bad form (in general) for a variable to change types as it makes code trickier to reason about. In your code `field_list_filter_class` changes from `str` to a class, where as in my version `field_list_filter_class` is always a class. This also keeps the name of the variable accurate, so if in the future extra code gets inserted before the `isinstance` check no one has to scratch their head wondering why a variable labeling itself as a class is being checked to see if it's a `str` (especially since their initial thought may be that it's a dotted path to a class, but, nope...).

I will concede that `second_param` is not a great variable name. 😉

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -986,6 +986,19 @@ subclass::
+          class BookAdmin(admin.ModelAdmin):
+              list_filter = (
+                  ('author__name', 'Author Name'),
+              )
```
It strikes me that using a string might actually be more useful that giving the `FieldListFilter` subclass directly. As such, I'd be inclined to put this example first. 

Something like: 

1. End `a tuple...` paragraph with `...for example, with a string title::`
2. This `BookAdmin` code example. 
3. Sentence saying, `...useful for fields on a related model...`
4. "...Or with a `FieldListFilter` subclass":
5. Existing examples and discussion. 
6. The `versionadded` note below.

carltongibson commented on django/contrib/admin/filters.py:
```diff
@@ -153,26 +153,28 @@ def register(cls, test, list_filter_class, take_priority=False):
-                return list_filter_class(field, request, params, model, model_admin, field_path=field_path)
+                return list_filter_class(
+                    field, request, params, model, model_admin, field_path=field_path, title=title
+                )
```
This should just fit on one line I think. (It's exactly 120 characters. 🙂)

dsanders11 replied:
Flake8 check is for 119 characters, so it fails the check at 120.

dsanders11 commented:
@carltongibson, didn't notice your comments until just now, I'll look at addressing them and making the separate PR in the next day or two.

dsanders11 commented:
@carltongibson, addressed your comment on the docs section, tried your 120 character suggestion but it failed the Flake8 build, and left a response to your other comment.

Pulled the checks `issubclass` change into #9844, but now this PR is in the awkward spot where it should be using that change, so I either way for that PR to be accepted and change this on to use it, or this code will need a minor tweak once both PRs are merged.

carltongibson commented:
Hi @dsanders11. 

* The docs changes look good I think. The docs build is failing. You'll need to fix up that. 
* You should be able to complete #9844 and then rebase this off of that. (I guess that's the simplest order.)

felixxm commented:
Updated in #10788.
